scripts/linking.py

```diff
@@ -481,30 +481,43 @@ def get_mention_text_and_ids_by_doc(data: List[data_util.MedMentionExample],
 def get_predicted_mention_candidates_and_types(span,
                                                ner_entities,
                                                filtered_batch_candidate_neighbor_ids,
-                                               predicted_mention_types):
-    candidates = {}
-    mention_types = set()
+                                               predicted_mention_types,
+                                               use_soft_matching):
+    """
+    This function returns three lists, candidates, mention types, and mention spans. These lists will be the same length and have
+    length equal to the number of predicted entities that overlap with the input gold entity. When not using soft mentions,
+    this length will be equal to one, as only one predicted entity can exactly match a gold entity.
+    """
+    candidates = []
+    mention_types = []
+    mention_spans = []
 
     if span is not None:
         for j, predicted_entity in enumerate(ner_entities):
-            overlaps = False
-            # gold span within spacy span
-            if span.start_char >= predicted_entity.start_char and span.end_char <= predicted_entity.end_char \
-                and predicted_entity != span:
-                overlaps = True
-            # spacy span within gold span
-            if predicted_entity.start_char >= span.start_char and predicted_entity.end_char <= span.end_char \
-                and predicted_entity != span:
-                overlaps = True
-            # endpoint overlap between gold span and spacy span
-            if predicted_entity.start_char <= span.start_char and predicted_entity.end_char >= span.start_char \
-                or predicted_entity.start_char <= span.end_char and predicted_entity.end_char >= span.end_char:
-                overlaps = True
-            if overlaps:
-                candidates.update(filtered_batch_candidate_neighbor_ids[j])
-                mention_types.update(predicted_mention_types[j]) 
-
-    return candidates, mention_types
+            if not use_soft_matching and span == predicted_entity:
+                candidates = filtered_batch_candidate_neighbor_ids[j]
+                mention_types = predicted_mention_types[j]
+                break
+            elif use_soft_matching:
+                overlaps = False
+                # gold span within spacy span
+                if span.start_char >= predicted_entity.start_char and span.end_char <= predicted_entity.end_char \
+                    and predicted_entity != span:
```

Comment:
Doesn't this logic miss out when the predicted span is exactly equal to the gold span, due to the `predicted_entity != span` clause? I think this case should be included.

Reply:
No, the third if statement handles that exactly equal case, although that is quite unclear. I will make it clearer

Reply:
Oh my bad

FYI I think this logic can be replaced with:

```python
    # Gold span starts inside predicted span.
if (span.start_char >= predicted_entity.start_char <= span.end_char
        # Predicted span starts inside gold span.
        or predicted_entity.start_char >= span.start_char <= predicted_entity.end_char):
	# do the overlap stuff.

```

Reply:
yeah, thats true. Iz pointed that out earlier, I'll just change it to that logic.

Reply:
I don't think this is working. I tried it and it gives the same result of the exact match. 
I think the overlapping logic needs to be:
```i0 = max(span_from_doc.start_char, predicted_entity.start_char)
i1 = min(span_from_doc.end_char, predicted_entity.end_char)
if i0 <= i1:
    overlap
else: 
   no overlap```

Reply:
I take it back. This is correct